[ie/instagram] Add `app_id` extractor-arg

chrisellsworth opened the pull request:
<!--
    **IMPORTANT**: PRs without the template will be CLOSED
    
    Due to the high volume of pull requests, it may be a while before your PR is reviewed.
    Please try to keep your pull request focused on a single bugfix or new feature.
    Pull requests with a vast scope and/or very large diff will take much longer to review.
    It is recommended for new contributors to stick to smaller pull requests, so you can receive much more immediate feedback as you familiarize yourself with the codebase.

    PLEASE AVOID FORCE-PUSHING after opening a PR, as it makes reviewing more difficult.
-->

### Description of your *pull request* and other information

Would like to add the ability to override the hardcoded Instagram headers as the response may vary depending on the values.

For some responses, media_id is already formatted as we need so there is no need to parse it out.

<details open><summary>Template</summary> <!-- OPEN is intentional -->

<!--
    # PLEASE FOLLOW THE GUIDE BELOW

    - You will be asked some questions, please read them **carefully** and answer honestly
    - Put an `x` into all the boxes `[ ]` relevant to your *pull request* (like [x])
    - Use *Preview* tab to see what your *pull request* will actually look like
-->

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check those that apply and remove the others:
- [X] I am the original author of the code in this PR, and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of the code in this PR, but it is in the public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*? Check those that apply and remove the others:
- [X] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

bashonly commented:
> Would like to add the ability to override the hardcoded Instagram headers as the response may vary depending on the values.

Please elaborate on this. Is there something the extractor should be doing differently?

bashonly commented on yt_dlp/extractor/instagram.py:
```diff
@@ -28,7 +28,8 @@
 
 def _pk_to_id(media_id):
     """Source: https://stackoverflow.com/questions/24437823/getting-instagram-post-url-from-media-id"""
-    return encode_base_n(int(media_id.split('_')[0]), table=_ENCODING_CHARS)
+    pk = int(media_id.split('_')[0]) if type(media_id) is str else media_id
+    return encode_base_n(pk, table=_ENCODING_CHARS)
```
Not a big fan of doing the type checking/conversion inside the method. Can you give an example where we pass `media_id` as an `int`?

chrisellsworth replied:
Will elaborate below!

bashonly replied:
Let's just do this then
```py
    pk = int(str(media_id).split('_')[0])
    return encode_base_n(pk, table=_ENCODING_CHARS)

```

bashonly commented on yt_dlp/extractor/instagram.py:
```diff
@@ -42,13 +43,23 @@ class InstagramBaseIE(InfoExtractor):
+        }
+
+        headers = {}
+        for k, v in base_headers.items():
+            if k in self.get_param('http_headers'):
+                headers[k] = self.get_param('http_headers')[k]
```
this will overwrite the `Accept` header every time

chrisellsworth replied:
Ah, yes, I see there's an inherent value of `'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8'` (though it does not seem to affect the response).

chrisellsworth commented:
Thanks for looking!

> > Would like to add the ability to override the hardcoded Instagram headers as the response may vary depending on the values.
> 
> Please elaborate on this. Is there something the extractor should be doing differently?

I'm trying to be generic about the approach but the core thing I'm looking to solve for is that the extractor is hardcoded to a specific client app ID. The encodings that IG provides may vary between client apps, so if you want a specific encoding you may need to behave as a different app. For example take the following video:

https://www.instagram.com/p/DFQPbvbSnIj/

The default output provides the following encodings:

```
❯ ./yt-dlp.sh https://www.instagram.com/p/DFQPbvbSnIj/ -F
[Instagram] Extracting URL: https://www.instagram.com/p/DFQPbvbSnIj/
[Instagram] DFQPbvbSnIj: Setting up session
[Instagram] DFQPbvbSnIj: Downloading JSON metadata
[info] Available formats for DFQPbvbSnIj:
ID                      EXT RESOLUTION │   FILESIZE  TBR PROTO │ VCODEC         VBR ACODEC     ABR ASR MORE INFO
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
dash-1173001884774598ad m4a audio only │ ≈120.16KiB  48k https │ audio only         mp4a.40.5  48k 44k DASH audio, m4a_dash
dash-883943686999326vd  mp4 538x736    │ ≈281.55KiB 114k https │ vp09.00.30.08 114k video only         DASH video, mp4_dash
dash-1137159621274847v  mp4 720x982    │ ≈601.64KiB 243k https │ vp09.00.31.08 243k video only         DASH video, mp4_dash
dash-920028790341606v   mp4 720x982    │ ≈918.36KiB 371k https │ vp09.00.31.08 371k video only         DASH video, mp4_dash
dash-602121562538092v   mp4 720x982    │ ≈  1.25MiB 517k https │ vp09.00.31.08 517k video only         DASH video, mp4_dash
5                       mp4 1080x1474  │                 https │ unknown            unknown
dash-1264987091460773v  mp4 1080x1474  │ ≈  1.83MiB 756k https │ vp09.00.40.08 756k video only         DASH video, mp4_dash
```

Using values derived from [this library](https://github.com/zweed4u/IG-Wrapper/blob/master/instagram_api.py#L29) we can get other encodings:

```
❯ ./yt-dlp.sh https://www.instagram.com/p/DFQPbvbSnIj/ --cookies-from-browser chrome --add-headers "X-IG-App-ID: 124024574287414" --add-headers "User-Agent: Instagram 339.0.0.12.95 (iPhone16,1; iOS 18_2; en_US; en-US; scale=3.00; gamut=normal; 1179x2556) AppleWebKit/420+" -F
[Instagram] Extracting URL: https://www.instagram.com/p/DFQPbvbSnIj/
Extracting cookies from chrome
Extracted 1797 cookies from chrome
[Instagram] DFQPbvbSnIj: Downloading video info
[info] Available formats for DFQPbvbSnIj:
ID                      EXT RESOLUTION │   FILESIZE  TBR PROTO │ VCODEC         VBR ACODEC     ABR ASR MORE INFO
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
dash-538031955954686ad  m4a audio only │ ≈ 79.07KiB  32k https │ audio only         mp4a.40.42 32k 44k DASH audio, m4a_dash
dash-1152773409785806a  m4a audio only │ ≈106.63KiB  43k https │ audio only         mp4a.40.42 43k 44k DASH audio, m4a_dash
dash-1013716917451013a  m4a audio only │ ≈145.13KiB  59k https │ audio only         mp4a.40.42 59k 44k DASH audio, m4a_dash
dash-1729790717588298vd mp4 538x736    │ ≈298.54KiB 120k https │ av01.0.04M.08 120k video only         DASH video, mp4_dash
921114300231397v-0      mp4 720x982    │                 https │ unknown            unknown
921114300231397v-1      mp4 720x982    │                 https │ unknown            unknown
921114300231397v-2      mp4 720x982    │                 https │ unknown            unknown
dash-1157971919073534v  mp4 720x982    │ ≈578.80KiB 234k https │ av01.0.05M.08 234k video only         DASH video, mp4_dash
dash-586424914255575v   mp4 720x982    │ ≈794.39KiB 321k https │ av01.0.05M.08 321k video only         DASH video, mp4_dash
dash-468027929714021v   mp4 720x982    │ ≈  1.02MiB 420k https │ av01.0.05M.08 420k video only         DASH video, mp4_dash
dash-628665002946797v   mp4 720x982    │ ≈  1.27MiB 524k https │ av01.0.05M.08 524k video only         DASH video, mp4_dash
dash-2864359213724424v  mp4 1080x1474  │ ≈  1.77MiB 731k https │ av01.0.08M.08 731k video only         DASH video, mp4_dash
```

But that second request only works with the type fix. Please observe that without it, we get this error: 

```
ERROR: 'int' object has no attribute 'split'
Traceback (most recent call last):
  File "~/yt-dlp/yt_dlp/YoutubeDL.py", line 1637, in wrapper
    return func(self, *args, **kwargs)
  File "~/yt-dlp/yt_dlp/YoutubeDL.py", line 1772, in __extract_info
    ie_result = ie.extract(url)
  File "~/yt-dlp/yt_dlp/extractor/common.py", line 747, in extract
    ie_result = self._real_extract(url)
  File "~/yt-dlp/yt_dlp/extractor/instagram.py", line 408, in _real_extract
    return self._extract_product(media)
           ~~~~~~~~~~~~~~~~~~~~~^^^^^^^
  File "~/yt-dlp/yt_dlp/extractor/instagram.py", line 187, in _extract_product
    '__post_extractor': self.extract_comments(_pk_to_id(product_info.get('pk'))),
                                              ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/yt-dlp/yt_dlp/extractor/instagram.py", line 31, in _pk_to_id
    return encode_base_n(int(media_id.split('_')[0]), table=_ENCODING_CHARS)
                             ^^^^^^^^^^^^^^
AttributeError: 'int' object has no attribute 'split'
```

Looking at the response, it appears to be because the default response has a `product_info` containing this:

`'pk': '3519386524481910596'` (a string)

Whereas the response with the overridden App ID contains this:

`'pk': 3519386524481910596` (already an int)

----

So, I can propose a few options if this is something you're interested in supporting:

(1) for the headers, add a new CLI parameter just for the value of App ID, similar to some of the YouTube-specific parameters?
(2) for the pk, given this only seems to break for `extract_comments`, maybe raise a warning rather than an error ("comments not supported for this request" or similar) so that it's still possible to get to the video encoding?

chrisellsworth commented:
Thanks for looking!

> > Would like to add the ability to override the hardcoded Instagram headers as the response may vary depending on the values.
> 
> Please elaborate on this. Is there something the extractor should be doing differently?

I'm trying to be generic about the approach but the core thing I'm looking to solve for is that the extractor is hardcoded to a specific client app ID. The encodings that IG provides may vary between client apps, so if you want a specific encoding you may need to behave as a different app. For example take the following video:

https://www.instagram.com/p/DFQPbvbSnIj/

The default output provides the following encodings:

```
❯ ./yt-dlp.sh https://www.instagram.com/p/DFQPbvbSnIj/ -F
[Instagram] Extracting URL: https://www.instagram.com/p/DFQPbvbSnIj/
[Instagram] DFQPbvbSnIj: Setting up session
[Instagram] DFQPbvbSnIj: Downloading JSON metadata
[info] Available formats for DFQPbvbSnIj:
ID                      EXT RESOLUTION │   FILESIZE  TBR PROTO │ VCODEC         VBR ACODEC     ABR ASR MORE INFO
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
dash-1173001884774598ad m4a audio only │ ≈120.16KiB  48k https │ audio only         mp4a.40.5  48k 44k DASH audio, m4a_dash
dash-883943686999326vd  mp4 538x736    │ ≈281.55KiB 114k https │ vp09.00.30.08 114k video only         DASH video, mp4_dash
dash-1137159621274847v  mp4 720x982    │ ≈601.64KiB 243k https │ vp09.00.31.08 243k video only         DASH video, mp4_dash
dash-920028790341606v   mp4 720x982    │ ≈918.36KiB 371k https │ vp09.00.31.08 371k video only         DASH video, mp4_dash
dash-602121562538092v   mp4 720x982    │ ≈  1.25MiB 517k https │ vp09.00.31.08 517k video only         DASH video, mp4_dash
5                       mp4 1080x1474  │                 https │ unknown            unknown
dash-1264987091460773v  mp4 1080x1474  │ ≈  1.83MiB 756k https │ vp09.00.40.08 756k video only         DASH video, mp4_dash
```

Using values derived from [this library](https://github.com/zweed4u/IG-Wrapper/blob/master/instagram_api.py#L29) we can get other encodings:

```
❯ ./yt-dlp.sh https://www.instagram.com/p/DFQPbvbSnIj/ --cookies-from-browser chrome --add-headers "X-IG-App-ID: 124024574287414" --add-headers "User-Agent: Instagram 339.0.0.12.95 (iPhone16,1; iOS 18_2; en_US; en-US; scale=3.00; gamut=normal; 1179x2556) AppleWebKit/420+" -F
[Instagram] Extracting URL: https://www.instagram.com/p/DFQPbvbSnIj/
Extracting cookies from chrome
Extracted 1797 cookies from chrome
[Instagram] DFQPbvbSnIj: Downloading video info
[info] Available formats for DFQPbvbSnIj:
ID                      EXT RESOLUTION │   FILESIZE  TBR PROTO │ VCODEC         VBR ACODEC     ABR ASR MORE INFO
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
dash-538031955954686ad  m4a audio only │ ≈ 79.07KiB  32k https │ audio only         mp4a.40.42 32k 44k DASH audio, m4a_dash
dash-1152773409785806a  m4a audio only │ ≈106.63KiB  43k https │ audio only         mp4a.40.42 43k 44k DASH audio, m4a_dash
dash-1013716917451013a  m4a audio only │ ≈145.13KiB  59k https │ audio only         mp4a.40.42 59k 44k DASH audio, m4a_dash
dash-1729790717588298vd mp4 538x736    │ ≈298.54KiB 120k https │ av01.0.04M.08 120k video only         DASH video, mp4_dash
921114300231397v-0      mp4 720x982    │                 https │ unknown            unknown
921114300231397v-1      mp4 720x982    │                 https │ unknown            unknown
921114300231397v-2      mp4 720x982    │                 https │ unknown            unknown
dash-1157971919073534v  mp4 720x982    │ ≈578.80KiB 234k https │ av01.0.05M.08 234k video only         DASH video, mp4_dash
dash-586424914255575v   mp4 720x982    │ ≈794.39KiB 321k https │ av01.0.05M.08 321k video only         DASH video, mp4_dash
dash-468027929714021v   mp4 720x982    │ ≈  1.02MiB 420k https │ av01.0.05M.08 420k video only         DASH video, mp4_dash
dash-628665002946797v   mp4 720x982    │ ≈  1.27MiB 524k https │ av01.0.05M.08 524k video only         DASH video, mp4_dash
dash-2864359213724424v  mp4 1080x1474  │ ≈  1.77MiB 731k https │ av01.0.08M.08 731k video only         DASH video, mp4_dash
```

But that second request only works with the type fix. Please observe that without it, we get this error: 

```
ERROR: 'int' object has no attribute 'split'
Traceback (most recent call last):
  File "~/yt-dlp/yt_dlp/YoutubeDL.py", line 1637, in wrapper
    return func(self, *args, **kwargs)
  File "~/yt-dlp/yt_dlp/YoutubeDL.py", line 1772, in __extract_info
    ie_result = ie.extract(url)
  File "~/yt-dlp/yt_dlp/extractor/common.py", line 747, in extract
    ie_result = self._real_extract(url)
  File "~/yt-dlp/yt_dlp/extractor/instagram.py", line 408, in _real_extract
    return self._extract_product(media)
           ~~~~~~~~~~~~~~~~~~~~~^^^^^^^
  File "~/yt-dlp/yt_dlp/extractor/instagram.py", line 187, in _extract_product
    '__post_extractor': self.extract_comments(_pk_to_id(product_info.get('pk'))),
                                              ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/yt-dlp/yt_dlp/extractor/instagram.py", line 31, in _pk_to_id
    return encode_base_n(int(media_id.split('_')[0]), table=_ENCODING_CHARS)
                             ^^^^^^^^^^^^^^
AttributeError: 'int' object has no attribute 'split'
```

Looking at the response, it appears to be because the default response has a `product_info` containing this:

`'pk': '3519386524481910596'` (a string)

Whereas the response with the overridden App ID contains this:

`'pk': 3519386524481910596` (already an int)

----

So, I can propose a few options if this is something you're interested in supporting:

(1) for the headers, add a new CLI parameter just for the value of App ID, similar to some of the YouTube-specific parameters?
(2) for the pk, given this only seems to break for `extract_comments`, maybe raise a warning rather than an error ("comments not supported for this request" or similar) so that it's still possible to get to the video encoding?

bashonly commented on yt_dlp/extractor/instagram.py:
```diff
@@ -28,7 +28,8 @@
+    return encode_base_n(pk, table=_ENCODING_CHARS)
 
 
 def _id_to_pk(shortcode):
```
Pretty sure `_id_to_pk` actually needs to be this or else it will convert incorrectly for private shortcodes from years ago
```diff
 def _id_to_pk(shortcode):
-    """Covert a shortcode to a numeric value"""
+    """Convert a shortcode to a numeric value"""
-    return decode_base_n(shortcode[:11], table=_ENCODING_CHARS)
+    return decode_base_n(shortcode[:-28], table=_ENCODING_CHARS)
```

bashonly replied:
Do you not want to do this in this PR?

chrisellsworth replied:
Happy to add it here too, was not sure if it was relevant to this PR or just an observation

bashonly reviewed (changes requested):
I'd prefer app ID to be an extractor-arg (`self._configuration_arg`)

chrisellsworth commented:
Thanks! Addressed these concerns. Tested with:

```
❯ ./yt-dlp.sh https://www.instagram.com/p/DFQPbvbSnIj/ --cookies-from-browser chrome --extractor-args "instagram:app_id=124024574287414" --add-headers "User-Agent: Instagram 339.0.0.12.95 (iPhone16,1; iOS 18_2; en_US; en-US; scale=3.00; gamut=normal; 1179x2556) AppleWebKit/420+" -F --print-traffic
```

bashonly commented on yt_dlp/extractor/instagram.py:
```diff
@@ -42,13 +43,15 @@ class InstagramBaseIE(InfoExtractor):
+
+    def _api_headers(self):
+        return {
+            'X-IG-App-ID': self._configuration_arg('app_id', '936619743392459')[0],
```
- default value needs to be a list with 1 string element or else we are slicing the string
- this is in the baseclass, so let's lock in the `ie_key` (or else we'd have to document `instagramstory:` etc)
```suggestion
            'X-IG-App-ID': self._configuration_arg('app_id', ['936619743392459'], ie_key=InstagramIE)[0],
```

Also I had a thought about keeping a map of known/common app IDs and allowing the user to pass `web`, `ios`, etc... but if we only know those 2 then I don't think it's worth it. What do you think?

chrisellsworth replied:
Agree, the extent, mapping, and stability of app IDs is not apparent at this point and not clear if there's a use case for overriding by anyone other than myself, so this seems sufficient for now. Perhaps some more exhaustive research could result in a future PR that does what you suggest.

bashonly commented on yt_dlp/extractor/instagram.py:
```diff
@@ -42,13 +43,15 @@ class InstagramBaseIE(InfoExtractor):
-        'Accept': '*/*',
-    }
+
+    def _api_headers(self):
```
I'd prefer to do this, and then remove all the `()`s below
```suggestion
    @property
    def _api_headers(self):
```

chrisellsworth commented:
Made updates per feedback; thanks!

bashonly commented on README.md:
```diff
@@ -1811,6 +1811,9 @@ The following extractors use this feature:
 * `dr`: dynamic range to ignore - one or more of `sdr`, `hdr10`, `dv`
 
+#### instagram
+* `app_id`: the value of the `X-IG-App-ID` header (default is `936619743392459`)
```
```suggestion
* `app_id`: The value of the `X-IG-App-ID` header used for API requests. Default is the web app ID, `936619743392459`
```

bashonly reviewed (approved):
Looks good. In the future, please do not force push
